### DISCUSSION TITLE
Architecture of modules when building apps at scale

### DISCUSSION THREAD

**Discussion by matiasvillaverde** [MobileNativeFoundation/discussions#137] [General]
I’m curious to hear the approach other apps have taken to architect the relationship between modules. I am mostly interested in problems that occur when building apps at scale.

1. Which types of modules do you have? What is the purpose of each?
2. How do you share code between modules?
3. How do you manage ownership of modules?
4. How do you make sure developers can work on focused modules and avoid compiling the whole project as part of their daily work?

**Comment by matiasvillaverde** [comment]
Sharing some insights of how we work at SoundCloud: 

1.  We have around 35 modules and the scratch build time for the main target is around 5 min 30 sec. Some of our modules are: 

`Main Target` We use the main target as a glue between features. Therefore, navigation, factory logic, and dependency injection goes into the main target. Because the application was initially built as a monolithic, we also have a lot of legacy code.

Down the stack we have a module called `Features` where we declare abstractions. Product modules can use these abstractions and by them we share code between product features. The main target will then instantiate the implementation of that abstractions and pass them down in the initialization of the product feature according to the dependency inversion principle.

`Core` This module contains the essential building blocks of the iOS application, it includes both implementations and protocols related to the following topics: Networking, Keychain access, Urn, Operations, Logging and some legacy libraries.

**Horizontal modules:** we have some modules that can be directly imported from feature modules. Some examples are: 

- `CoreUI` This module contains the base building blocks for UI components. This can be the definitions of colors, sizes and fonts.
- `UIComponets` In the module we have the building blocks for the UI layer. Here we define buttons, labels, lists, etc. The product modules will use these building blocks to create UI. 
- `SCData` in this module we encapsulate the access to data behind repositories. Here goes most of the caching logic.

**Product Modules:** each product feature is splitted into two modules: Foundation and Kit. In the foundation module goes the presentation and business logic. On the Kit module goes the UI. Some examples are `UserProfile`, `TrackUpload`, `Collection`, etc. 


2. Core, features and horizontal modules can be imported directly from Product modules. This makes it easy to share code, but also it has an impact on building times. Given that engineers must compile all the horizontal modules when compiling a product module.  On the other hand, product Modules share code between them via the Features module following the dependency inversion principle. We declare abstractions in Features, and the main target will instantiate the implementations.

3. We currently have a spreadsheet with teams owning certain modules. As well, we have an implicit knowledge of who owns what. Unfortunately, this method is not scalable because we go through many reorganizations and the spreadsheet is out of date. 

4. Each product module has an example application. Engineers need to compile only their product module + horizontal modules + features + Core to work isolated from the rest of the app. Once the feature is ready, they integrate it into the main target and only then they compile the whole project.

**Comment by sberrevoets** [comment]
There was some more discussion on this in #72 as well

**Comment by mattrobmattrob** [comment]
This [very recent blog post from AirBnb is also relevant](https://medium.com/airbnb-engineering/designing-for-productivity-in-a-large-scale-ios-application-9376a430a0bf):

>To address the lack of hierarchy in our code, and therefore lack of discoverability, we have designed an organizational structure for our first-party code. Modules are organized into semantically meaningful groups, called module types.
>
>We have written precise documentation for our module types. Since the concept of module types is so fundamental to the way iOS developers work at Airbnb, this documentation is hosted on our internal developer portal and managed in source control. We summarize each module type in just a few paragraphs, explaining the purpose of the module type and the types of code it was designed to support.
>
>We considered both application programming and build system best practices when designing this architecture. Each module type has a strict set of visibility rules. These visibility rules define the allowed dependencies between modules of that type. An individual module may tighten its visibility, a technique used by some larger teams who enjoy the benefits of modularity and want to avoid unexpected inbound dependencies on their modules. An individual module cannot expand its visibility beyond the limits imposed by its module type.
